src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -96,37 +102,86 @@ def get_packages_to_cover(
 
 @rule
 async def setup_pytest_for_target(
-    adaptor_with_origin: PythonTestsAdaptorWithOrigin, pytest: PyTest, test_options: TestOptions,
+    adaptor_with_origin: PythonTestsAdaptorWithOrigin,
+    pytest: PyTest,
+    test_options: TestOptions,
+    python_setup: PythonSetup,
 ) -> TestTargetSetup:
-    adaptor = adaptor_with_origin.adaptor
     # TODO: Rather than consuming the TestOptions subsystem, the TestRunner should pass on coverage
     # configuration via #7490.
-    transitive_hydrated_targets = await Get[TransitiveHydratedTargets](
-        Addresses((adaptor.address,))
-    )
+
+    adaptor = adaptor_with_origin.adaptor
+    test_addresses = Addresses((adaptor.address,))
+
+    # TODO(John Sirois): PexInterpreterConstraints are gathered in the same way by the
+    #  `create_pex_from_target_closure` rule, factor up.
+    transitive_hydrated_targets = await Get[TransitiveHydratedTargets](Addresses, test_addresses)
     all_targets = transitive_hydrated_targets.closure
+    all_target_adaptors = [t.adaptor for t in all_targets]
+    interpreter_constraints = PexInterpreterConstraints.create_from_adaptors(
+        adaptors=all_target_adaptors, python_setup=python_setup
+    )
 
-    resolved_requirements_pex = await Get[Pex](
+    # Ensure all pexes we merge via PEX_PATH to form the test runner use the interpreter constraints
+    # of the tests. This is handled by CreatePexFromTargetClosure, but we must pass this through for
+    # CreatePex requests.
+    create_pex = functools.partial(CreatePex, interpreter_constraints=interpreter_constraints)
+
+    # NB: We set `--not-zip-safe` because Pytest plugin discovery, which uses
+    # `importlib_metadata` and thus `zipp`, does not play nicely when doing import magic
+    # directly
+    # from zip files. `zipp` has pathologically bad behavior with large zipfiles.
+    # TODO: this does have a performance cost as the pex must now be expanded to disk. Long
+    #  term,
+    # it would be better to fix Zipp (whose fix would then need to be used by importlib_metadata
+    # and then by Pytest). See https://github.com/jaraco/zipp/pull/26.
```

Comment:
Nit: formatting of this got messed up when copying over. (I wish Black could safely format these for us.)

Reply:
Fixed.